src/AspNet.Security.OAuth.Xiaomi/XiaomiAuthenticationHandler.cs

```diff
@@ -0,0 +1,162 @@
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.UserProfileErrorAsync(Logger, response, Context.RequestAborted);
+            throw new HttpRequestException("An error occurred while retrieving user information.");
+        }
+
+        var json = await response.Content.ReadAsStringAsync(Context.RequestAborted);
+        using var payload = JsonDocument.Parse(json);
+        var rootElement = payload.RootElement;
+
+        if (!rootElement.TryGetProperty("data", out JsonElement dataElement))
+        {
+            var errorCode = rootElement.GetString("code")!;
+            throw new Exception($"An error (Code: {errorCode}) occurred while retrieving user information.");
+        }
+
+        identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, dataElement.GetString("unionId")!, ClaimValueTypes.String, Options.ClaimsIssuer));
+
+        var principal = new ClaimsPrincipal(identity);
+        var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, dataElement);
+        context.RunClaimActions();
+
+        await Events.CreatingTicket(context);
+        return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
+    }
+
+    protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+    {
+        var tokenRequestParameters = new Dictionary<string, string?>()
+        {
+            ["client_id"] = Options.ClientId,
+            ["client_secret"] = Options.ClientSecret,
+            ["code"] = context.Code,
+            ["redirect_uri"] = context.RedirectUri,
+            ["grant_type"] = "authorization_code",
+        };
+
+        // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+        if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier))
+        {
+            tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+            context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+        }
+
+        using var request = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        request.Content = new FormUrlEncodedContent(tokenRequestParameters);
+
+        using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.ExchangeCodeErrorAsync(Logger, response, Context.RequestAborted);
+            return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
+        }
+
+        var json = await response.Content.ReadAsStringAsync(Context.RequestAborted);
+
+        var payload = JsonDocument.Parse(json);
+
+        var errorCode = payload.RootElement.GetString("errcode");
```

Comment:
I see in the docs (https://dev.mi.com/console/doc/detail?pId=707) that errors returned from the token endpoint are expected to include the standard `error` parameter. Are we sure `errcode` is really returned?

Reply:
Sorry, this is last year's code, I didn't revalidate, this should be I accidentally wrote the same code as the defined parameter name, I now try to change the request parameters to verify, it always returns an error with status code 400 (Use get or post requests), so I can't be sure if it returns errcode, and I think it may not return an error message with status code 200 after it fails the request, I now modify it to error as per the docs.

![65ab7f466c1c682f8aad2f9d036676a](https://user-images.githubusercontent.com/35516482/234448610-02da0ea4-78a2-4737-9c97-8b7d75b1f324.png)